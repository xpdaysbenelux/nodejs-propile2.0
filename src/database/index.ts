export * from './entities';
export * from './repositories';
export * from './database.module';
