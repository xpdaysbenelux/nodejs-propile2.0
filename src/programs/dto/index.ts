export * from './create-program.dto';
export * from './get-program.dto';
export * from './get-programs.dto';
export * from './update-program.dto';
