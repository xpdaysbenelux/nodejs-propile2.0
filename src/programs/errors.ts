import { BadRequestException } from '@nestjs/common';

export class ProgramTitleAlreadyInUse extends BadRequestException {
    constructor() {
        super(
            'A program with this title already exists.',
            'PROGRAM_TITLE_ALREADY_IN_USE',
        );
    }
}

export class ProgramDateMustBeBetweenConferenceDates extends BadRequestException {
    constructor() {
        super(
            'The program date must be one of or between the conference dates.',
            'PROGRAM_DATE_MUST_BE_BETWEEN_CONFERENCE_DATES',
        );
    }
}

export class StartEndTimeDatesMustBeSameAsProgramDate extends BadRequestException {
    constructor() {
        super(
            'The start and end time dates must be the same as the program date.',
            'START_END_TIME_DATES_MUST_BE_SAME_AS_PROGRAM_DATE',
        );
    }
}
