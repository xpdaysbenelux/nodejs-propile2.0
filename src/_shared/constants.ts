import { Permissions } from '../database';

export enum UserState {
    Registering = 'REGISTERING',
    Active = 'ACTIVE',
    Inactive = 'INACTIVE',
}

export enum SortDirection {
    Ascending = 'ASC',
    Descending = 'DESC',
}

export interface IUserSession {
    userId: string;
    email: string;
    state: UserState;
    firstName?: string;
    lastName?: string;
    permissions: Permissions;
}
