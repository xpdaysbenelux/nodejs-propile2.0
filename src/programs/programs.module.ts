import { Module } from '@nestjs/common';

import { ProgramController } from './programs.controller';
import { ProgramsService } from './programs.service';
import { ProgramsQueries } from './programs.queries';
import { DatabaseModule } from '../database';

@Module({
    imports: [DatabaseModule],
    controllers: [ProgramController],
    providers: [ProgramsService, ProgramsQueries],
})
export class ProgramsModule {}
