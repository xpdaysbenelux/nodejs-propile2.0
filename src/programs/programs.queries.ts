import { Injectable } from '@nestjs/common';

import { ProgramRepository } from '../database';
import { ProgramResponse } from './dto/get-program.dto';

const programFields = [
    'profram.id',
    'program.title',
    'program.date',
    'program.startTime',
    'program.endTime',
    'conference.id',
    'conference.name',
    'event.id',
    'event.spanRow',
    'event.startTime',
    'event.endTime',
    'event.room',
];

@Injectable()
export class ProgramQueries {
    constructor(private readonly programRepository: ProgramRepository) {}

    async getProgram(programId: string): Promise<ProgramResponse> {
        return this.programRepository
            .createQueryBuilder('program')
            .select(programFields)
            .where('program.id = :programId', { programId })
            .innerJoin('program.conference', 'conference')
            .leftJoin('program.events', 'event')
            .getOne();
    }
}
