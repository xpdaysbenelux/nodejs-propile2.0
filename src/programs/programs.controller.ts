import { Body, UseGuards, Controller, Post, Get, Param } from '@nestjs/common';
import { ApiTags } from '@nestjs/swagger';

import { ProgramsService } from './programs.service';
import { ProgramQueries } from './programs.queries';
import { CreateProgramRequest } from './dto/create-program.dto';
import { UserSession, RequiredPermissions } from '../_shared/decorators';
import { IUserSession } from '../_shared/constants';
import {
    AuthenticatedGuard,
    RequiredPermissionsGuard,
} from '../_shared/guards';
import { ProgramResponse, ProgramIdParam } from './dto/get-program.dto';

@UseGuards(AuthenticatedGuard)
@Controller('programs')
@ApiTags('programs')
export class ProgramController {
    constructor(
        private readonly programService: ProgramsService,
        private readonly programQueries: ProgramQueries,
    ) {}

    @RequiredPermissions({ programs: { view: true, edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Get(':programId')
    getConference(@Param() params: ProgramIdParam): Promise<ProgramResponse> {
        return this.programQueries.getProgram(params.programId);
    }

    @RequiredPermissions({ programs: { edit: true } })
    @UseGuards(RequiredPermissionsGuard)
    @Post()
    async createProgram(
        @Body() body: CreateProgramRequest,
        @UserSession() session: IUserSession,
    ): Promise<ProgramResponse> {
        const programId = await this.programService.createProgram(
            body,
            session,
        );
        return this.programQueries.getProgram(programId);
    }
}
