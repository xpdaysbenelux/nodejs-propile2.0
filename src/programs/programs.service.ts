import { Injectable } from '@nestjs/common';
import { parseISO, isWithinInterval, isSameDay } from 'date-fns';

import { ProgramRepository, Program, ConferenceRepository } from '../database';
import { CreateProgramRequest } from './dto/create-program.dto';
import { IUserSession } from '../_shared/constants';
import {
    ProgramTitleAlreadyInUse,
    ProgramDateMustBeBetweenConferenceDates,
    StartEndTimeDatesMustBeSameAsProgramDate,
    ProgramNotFoud,
} from './errors';
import { ConferenceNotFoud } from '../conferences/errors';
import { UpdateProgramRequest } from './dto';
import { Not } from 'typeorm';

@Injectable()
export class ProgramsService {
    constructor(
        private readonly programRepository: ProgramRepository,
        private readonly conferenceRepository: ConferenceRepository,
    ) {}

    async createProgram(
        body: CreateProgramRequest,
        session: IUserSession,
    ): Promise<string> {
        const { title, date, startTime, endTime, conferenceId } = body;
        const programDate = parseISO(date);
        const programStartTime = parseISO(startTime);
        const programEndTime = parseISO(endTime);

        const program = new Program();

        const existingProgram = await this.programRepository.findOne({
            title,
        });
        if (existingProgram) {
            throw new ProgramTitleAlreadyInUse();
        }
        program.title = title;

        const conference = await this.conferenceRepository.findOne({
            id: conferenceId,
        });
        if (!conference) {
            throw new ConferenceNotFoud();
        }
        program.conference = conference;

        // Check if the program date is one of or between the conference dates
        if (
            !isWithinInterval(programDate, {
                start: conference.startDate,
                end: conference.endDate,
            })
        ) {
            throw new ProgramDateMustBeBetweenConferenceDates();
        }
        program.date = programDate;

        // Check if the start & endTime dates are the same as the program date
        if (
            !isSameDay(programStartTime, programDate) ||
            !isSameDay(programEndTime, programDate)
        ) {
            throw new StartEndTimeDatesMustBeSameAsProgramDate();
        }
        program.startTime = programStartTime;
        program.endTime = programEndTime;

        program.createdBy = session.email;
        program.createdAt = new Date();

        await this.programRepository.save(program);

        return program.id;
    }

    async updateProgram(
        body: UpdateProgramRequest,
        programId: string,
        session: IUserSession,
    ): Promise<string> {
        const { title, date, startTime, endTime, conferenceId } = body;
        const programDate = parseISO(date);
        const programStartTime = parseISO(startTime);
        const programEndTime = parseISO(endTime);

        const existingProgram = await this.programRepository.findOne({
            where: { id: programId },
            relations: ['events'],
        });
        if (!existingProgram) {
            throw new ProgramNotFoud();
        }

        const existingProgramWithSameTitle = await this.programRepository.findOne(
            {
                where: { id: Not(programId), title },
            },
        );
        if (existingProgramWithSameTitle) {
            throw new ProgramTitleAlreadyInUse();
        }

        const conference = await this.conferenceRepository.findOne({
            id: conferenceId,
        });
        if (!conference) {
            throw new ConferenceNotFoud();
        }

        // Check if the program date is one of or between the conference dates
        if (
            !isWithinInterval(programDate, {
                start: conference.startDate,
                end: conference.endDate,
            })
        ) {
            throw new ProgramDateMustBeBetweenConferenceDates();
        }

        // Check if the start & endTime dates are the same as the program date
        if (
            !isSameDay(programStartTime, programDate) ||
            !isSameDay(programEndTime, programDate)
        ) {
            throw new StartEndTimeDatesMustBeSameAsProgramDate();
        }

        existingProgram.title = title;
        existingProgram.conference = conference;
        existingProgram.date = programDate;
        existingProgram.startTime = programStartTime;
        existingProgram.endTime = programEndTime;
        existingProgram.updatedBy = session.email;
        existingProgram.createdAt = new Date();

        await this.programRepository.save(existingProgram);

        return existingProgram.id;
    }

    async deleteProgram(programId: string): Promise<string> {
        const existingProgram = await this.programRepository.findOne({
            where: { id: programId },
            relations: ['events'],
        });
        if (!existingProgram) {
            throw new ProgramNotFoud();
        }

        await this.programRepository.delete(programId);
        return programId;
    }
}
