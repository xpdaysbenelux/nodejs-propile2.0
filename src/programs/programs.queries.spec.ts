import { Test } from '@nestjs/testing';
import * as faker from 'faker';

import { ProgramRepository, DatabaseModule } from '../database';
import { ProgramsQueries } from './programs.queries';
import { SortDirection } from '../_shared/constants';
import {
    GetProgramsRequestQuery,
    ProgramsSortColumns,
} from './dto/get-programs.dto';

describe('ProgramsQueries', () => {
    let programRepository: ProgramRepository;
    let programsQueries: ProgramsQueries;

    beforeAll(async () => {
        const module = await Test.createTestingModule({
            imports: [DatabaseModule],
            providers: [ProgramsQueries],
        }).compile();

        programsQueries = module.get(ProgramsQueries);
        programRepository = module.get(ProgramRepository);
    });

    afterAll(() => {
        programRepository.manager.connection.close();
    });

    describe('getProgram', () => {
        it('should return the requested program correctly', async () => {
            const result = await programsQueries.getProgram(
                '3185e221-73ca-4b5a-93a9-1cc2d0b5df76',
            );

            const { startTime, endTime, date, ...withoutDates } = result;
            expect(withoutDates).toMatchSnapshot();
        });

        it('should return nothing if the requested program does not exist', async () => {
            const result = await programsQueries.getProgram(
                faker.random.uuid(),
            );

            expect(result).toMatchSnapshot();
        });
    });

    describe('getPrograms', () => {
        const testQueries: GetProgramsRequestQuery[] = [
            {},
            { search: 'planning' },
            { search: 'program' },
            {
                sortBy: ProgramsSortColumns.Date,
                sortDirection: SortDirection.Ascending,
            },
        ];

        testQueries.forEach((query, index) => {
            it(`should return a paged list of programs: #${index}`, async () => {
                const result = await programsQueries.getPrograms(query);

                const withoutDatesList = result.data.map(entry => {
                    const { startTime, endTime, date, ...withoutDates } = entry;
                    return withoutDates;
                });
                expect({
                    data: withoutDatesList,
                    meta: result.meta,
                }).toMatchSnapshot();
            });
        });

        it('should return an empty list if no programs are found', async () => {
            const result = await programsQueries.getPrograms({
                search: 'jibbrishsearch',
            });
            expect(result).toMatchSnapshot();
        });
    });
});
