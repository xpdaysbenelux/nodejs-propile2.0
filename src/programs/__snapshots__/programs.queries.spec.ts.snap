// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgramsQueries getProgram should return nothing if the requested program does not exist 1`] = `undefined`;

exports[`ProgramsQueries getProgram should return the requested program correctly 1`] = `
Program {
  "conference": Conference {
    "id": "81bdc928-b92d-4111-91f9-3ef93e203e88",
    "name": "Conference 1",
  },
  "date": 2020-04-02T00:00:00.000Z,
  "endTime": 2020-04-02T18:00:00.000Z,
  "events": Array [],
  "id": "3185e221-73ca-4b5a-93a9-1cc2d0b5df76",
  "startTime": 2020-04-02T06:00:00.000Z,
  "title": "Conference planning day 1",
}
`;

exports[`ProgramsQueries getPrograms should return a paged list of programs: #0 1`] = `
Object {
  "data": Array [
    Program {
      "conference": Conference {
        "id": "81bdc928-b92d-4111-91f9-3ef93e203e88",
        "name": "Conference 1",
      },
      "date": 2020-04-02T00:00:00.000Z,
      "endTime": 2020-04-02T18:00:00.000Z,
      "events": Array [],
      "id": "3185e221-73ca-4b5a-93a9-1cc2d0b5df76",
      "startTime": 2020-04-02T06:00:00.000Z,
      "title": "Conference planning day 1",
    },
  ],
  "meta": Object {
    "count": 1,
    "skip": 0,
    "totalCount": 1,
  },
}
`;

exports[`ProgramsQueries getPrograms should return a paged list of programs: #1 1`] = `
Object {
  "data": Array [
    Program {
      "conference": Conference {
        "id": "81bdc928-b92d-4111-91f9-3ef93e203e88",
        "name": "Conference 1",
      },
      "date": 2020-04-02T00:00:00.000Z,
      "endTime": 2020-04-02T18:00:00.000Z,
      "events": Array [],
      "id": "3185e221-73ca-4b5a-93a9-1cc2d0b5df76",
      "startTime": 2020-04-02T06:00:00.000Z,
      "title": "Conference planning day 1",
    },
  ],
  "meta": Object {
    "count": 1,
    "skip": 0,
    "totalCount": 1,
  },
}
`;

exports[`ProgramsQueries getPrograms should return a paged list of programs: #2 1`] = `
Object {
  "data": Array [],
  "meta": Object {
    "count": 0,
    "skip": 0,
    "totalCount": 0,
  },
}
`;

exports[`ProgramsQueries getPrograms should return a paged list of programs: #3 1`] = `
Object {
  "data": Array [
    Program {
      "conference": Conference {
        "id": "81bdc928-b92d-4111-91f9-3ef93e203e88",
        "name": "Conference 1",
      },
      "date": 2020-04-02T00:00:00.000Z,
      "endTime": 2020-04-02T18:00:00.000Z,
      "events": Array [],
      "id": "3185e221-73ca-4b5a-93a9-1cc2d0b5df76",
      "startTime": 2020-04-02T06:00:00.000Z,
      "title": "Conference planning day 1",
    },
  ],
  "meta": Object {
    "count": 1,
    "skip": 0,
    "totalCount": 1,
  },
}
`;

exports[`ProgramsQueries getPrograms should return an empty list if no programs are found 1`] = `
Object {
  "data": Array [],
  "meta": Object {
    "count": 0,
    "skip": 0,
    "totalCount": 0,
  },
}
`;

exports[`ProgramsQueries getPrograms should return an empty list if no programs are found 2`] = `
Object {
  "data": Array [],
  "meta": Object {
    "count": 0,
    "skip": 0,
    "totalCount": 0,
  },
}
`;

exports[`ProgramsQueries getPrograms should return an empty list if no programs are found 3`] = `
Object {
  "data": Array [],
  "meta": Object {
    "count": 0,
    "skip": 0,
    "totalCount": 0,
  },
}
`;

exports[`ProgramsQueries getPrograms should return an empty list if no programs are found 4`] = `
Object {
  "data": Array [],
  "meta": Object {
    "count": 0,
    "skip": 0,
    "totalCount": 0,
  },
}
`;
